x-pack/plugins/fleet/server/services/metrics/fetch_agent_metrics.ts

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { ElasticsearchClient } from '@kbn/core-elasticsearch-server';
+import type { CoreSetup } from '@kbn/core/server';
+
+import { AGENTS_INDEX } from '../../../common';
+
+import type { AgentPerVersion, AgentUsage } from '../../collectors/agent_collectors';
+import { getAgentUsage } from '../../collectors/agent_collectors';
+import { getInternalClients } from '../../collectors/helpers';
+import { appContextService } from '../app_context';
+import { retryTransientEsErrors } from '../epm/elasticsearch/retry';
+
+export interface AgentMetrics {
+  agents: AgentUsage;
+  agents_per_version: AgentPerVersion[];
+  upgrading_step: UpgradingSteps;
+}
+
+export interface UpgradingSteps {
+  requested: number;
+  scheduled: number;
+  downloading: number;
+  extracting: number;
+  replacing: number;
+  restarting: number;
+  watching: number;
+  rollback: number;
+  failed: number;
+}
+
+export const fetchAgentMetrics = async (
+  core: CoreSetup,
+  abortController: AbortController
+): Promise<AgentMetrics | undefined> => {
+  const [soClient, esClient] = await getInternalClients(core);
+  if (!soClient || !esClient) {
+    return;
+  }
+  const usage = {
+    agents: await getAgentUsage(soClient, esClient),
+    agents_per_version: await getAgentsPerVersion(esClient, abortController),
+    upgrading_step: await getUpgradingSteps(esClient, abortController),
+  };
+  return usage;
+};
+
+export const getAgentsPerVersion = async (
+  esClient: ElasticsearchClient,
+  abortController: AbortController
+): Promise<AgentPerVersion[]> => {
+  try {
+    const response = await retryTransientEsErrors(() =>
+      esClient.search(
+        {
+          index: AGENTS_INDEX,
+          query: {
+            bool: {
+              filter: [
+                {
+                  term: {
+                    active: 'true',
+                  },
+                },
+              ],
+            },
+          },
+          size: 0,
+          aggs: {
+            versions: {
+              terms: { field: 'agent.version' },
+            },
+          },
+        },
+        { signal: abortController.signal }
+      )
+    );
+    return ((response?.aggregations?.versions as any).buckets ?? []).map((bucket: any) => ({
+      version: bucket.key,
+      count: bucket.doc_count,
+    }));
+  } catch (error) {
+    if (error.statusCode === 404) {
+      appContextService.getLogger().debug('Index .fleet-agents does not exist yet.');
```

Comment:
Is this worth logging at another level so we can see it in serverless dashboards, etc? Not sure how common this is or if it's helpful in production debugging to know when we're swallowing these errors.

Reply:
I want to avoid logging too much, as this task runs every minute. I think we can leave as debug for now and change later if needed. Probably we would realize anyway if there are no agents.